File: scripts/go/launch_admin_via_api.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"net/http"
	"net/http/cookiejar"
	"os"
	"regexp"
	"strings"
	"time"
)

type client struct {
	baseURL string
	http    *http.Client
}

type userProfile struct {
	Username string `json:"username"`
}

type dashboardSummary struct {
	ClusterCount      int `json:"clusterCount"`
	PodCount          int `json:"podCount"`
	DeploymentCount   int `json:"deploymentCount"`
	TodayReleaseCount int `json:"todayReleaseCount"`
}

type clusterSummary struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type clusterTestResponse struct {
	Success bool   `json:"success"`
	Version string `json:"version"`
	Error   string `json:"error"`
}

func newClient(baseURL string) (*client, error) {
	jar, err := cookiejar.New(nil)
	if err != nil {
		return nil, err
	}
	return &client{
		baseURL: strings.TrimRight(baseURL, "/"),
		http: &http.Client{
			Jar:     jar,
			Timeout: 20 * time.Second,
		},
	}, nil
}

func (c *client) do(method, path string, payload any, out any, allowedStatus ...int) (int, error) {
	allowed := map[int]struct{}{}
	for _, status := range allowedStatus {
		allowed[status] = struct{}{}
	}

	var body io.Reader
	if payload != nil {
		raw, err := json.Marshal(payload)
		if err != nil {
			return 0, err
		}
		body = bytes.NewBuffer(raw)
	}

	req, err := http.NewRequest(method, c.baseURL+path, body)
	if err != nil {
		return 0, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.http.Do(req)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()

	rawBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return resp.StatusCode, err
	}

	if out != nil && len(rawBody) > 0 {
		if err := json.Unmarshal(rawBody, out); err != nil {
			return resp.StatusCode, err
		}
	}

	if resp.StatusCode >= 400 {
		if _, ok := allowed[resp.StatusCode]; !ok {
			return resp.StatusCode, fmt.Errorf("http %d: %s", resp.StatusCode, string(rawBody))
		}
	}

	return resp.StatusCode, nil
}

func resolvePassword(password, passwordEnv string) (string, error) {
	if password != "" {
		return password, nil
	}
	if passwordEnv != "" {
		value := os.Getenv(passwordEnv)
		if value != "" {
			return value, nil
		}
		return "", fmt.Errorf("environment variable %s is empty", passwordEnv)
	}
	return "", errors.New("password is required (--password or --password-env)")
}

func retry(action string, maxRetries int, retryInterval time.Duration, fn func() error) error {
	for attempt := 1; attempt <= maxRetries; attempt++ {
		if err := fn(); err != nil {
			if attempt == maxRetries {
				return fmt.Errorf("%s failed after %d attempts: %w", action, attempt, err)
			}
			fmt.Fprintf(os.Stderr, "%s attempt %d/%d failed: %v; retrying in %s\n", action, attempt, maxRetries, err, retryInterval)
			time.Sleep(retryInterval)
			continue
		}
		return nil
	}
	return fmt.Errorf("%s failed", action)
}

func extractAPIServerURL(kubeconfig string) (string, error) {
	re := regexp.MustCompile(`(?m)^\s*server:\s*(\S+)\s*$`)
	matches := re.FindStringSubmatch(kubeconfig)
	if len(matches) < 2 {
		return "", errors.New("kubeconfig does not contain a server field")
	}
	return matches[1], nil
}

func main() {
	baseURL := flag.String("base-url", "http://127.0.0.1:3002", "admin base url")
	username := flag.String("username", "admin", "login username")
	password := flag.String("password", "", "login password")
	passwordEnv := flag.String("password-env", "", "environment variable name containing login password")

	upsertCluster := flag.Bool("upsert-cluster", false, "upsert cluster using /api/clusters then optional /api/clusters/{id}/test")
	clusterName := flag.String("cluster-name", "", "cluster unique name")
	clusterDisplayName := flag.String("cluster-display-name", "", "cluster display name")
	clusterDescription := flag.String("cluster-description", "Managed by Terraform API workflow", "cluster description used in upsert payload")
	kubeconfigPath := flag.String("kubeconfig", "", "path to kubeconfig file")
	apiServerURL := flag.String("api-server-url", "", "optional API server URL override; inferred from kubeconfig when empty")
	testCluster := flag.Bool("test-cluster", false, "run /api/clusters/{id}/test after upsert")

	maxRetries := flag.Int("max-retries", 30, "maximum retry count for login and optional connectivity tests")
	retryIntervalSeconds := flag.Int("retry-interval-seconds", 2, "retry interval in seconds")
	skipDashboard := flag.Bool("skip-dashboard", false, "skip /api/dashboard verification")
	flag.Parse()

	if *maxRetries <= 0 {
		fmt.Fprintln(os.Stderr, "ERROR: --max-retries must be > 0")
		os.Exit(1)
	}
	if *retryIntervalSeconds <= 0 {
		fmt.Fprintln(os.Stderr, "ERROR: --retry-interval-seconds must be > 0")
		os.Exit(1)
	}
	if *upsertCluster && (*clusterName == "" || *clusterDisplayName == "" || *kubeconfigPath == "") {
		fmt.Fprintln(os.Stderr, "ERROR: --upsert-cluster requires --cluster-name, --cluster-display-name and --kubeconfig")
		os.Exit(1)
	}

	resolvedPassword, err := resolvePassword(*password, *passwordEnv)
	if err != nil {
		fmt.Fprintln(os.Stderr, "ERROR:", err)
		os.Exit(1)
	}

	c, err := newClient(*baseURL)
	if err != nil {
		fmt.Fprintln(os.Stderr, "ERROR:", err)
		os.Exit(1)
	}

	retryInterval := time.Duration(*retryIntervalSeconds) * time.Second
	loginReq := map[string]string{"username": *username, "password": resolvedPassword}
	if err := retry("login", *maxRetries, retryInterval, func() error {
		_, err := c.do(http.MethodPost, "/api/auth/login", loginReq, nil)
		return err
	}); err != nil {
		fmt.Fprintln(os.Stderr, "ERROR:", err)
		os.Exit(1)
	}

	var me userProfile
	if _, err := c.do(http.MethodGet, "/api/auth/me", nil, &me); err != nil {
		fmt.Fprintln(os.Stderr, "ERROR:", err)
		os.Exit(1)
	}
	fmt.Println("login_user:", me.Username)

	if *upsertCluster {
		kubeconfigContentBytes, err := os.ReadFile(*kubeconfigPath)
		if err != nil {
			fmt.Fprintln(os.Stderr, "ERROR:", err)
			os.Exit(1)
		}
		kubeconfigContent := string(kubeconfigContentBytes)
		resolvedAPIServerURL := strings.TrimSpace(*apiServerURL)
		if resolvedAPIServerURL == "" {
			resolvedAPIServerURL, err = extractAPIServerURL(kubeconfigContent)
			if err != nil {
				fmt.Fprintln(os.Stderr, "ERROR:", err)
				os.Exit(1)
			}
		}

		var clusterList []clusterSummary
		if _, err := c.do(http.MethodGet, "/api/clusters", nil, &clusterList); err != nil {
			fmt.Fprintln(os.Stderr, "ERROR:", err)
			os.Exit(1)
		}

		clusterPayload := map[string]any{
			"displayName":   *clusterDisplayName,
			"apiServerUrl":  resolvedAPIServerURL,
			"authType":      "kubeconfig",
			"kubeconfig":    kubeconfigContent,
			"description":   *clusterDescription,
			"notifyEnabled": false,
		}

		clusterID := ""
		upsertResult := ""
		for _, item := range clusterList {
			if item.Name == *clusterName {
				clusterID = item.ID
				break
			}
		}

		if clusterID == "" {
			createPayload := map[string]any{
				"name": *clusterName,
			}
			for k, v := range clusterPayload {
				createPayload[k] = v
			}
			var created clusterSummary
			if _, err := c.do(http.MethodPost, "/api/clusters", createPayload, &created); err != nil {
				fmt.Fprintln(os.Stderr, "ERROR:", err)
				os.Exit(1)
			}
			if created.ID == "" {
				fmt.Fprintln(os.Stderr, "ERROR: cluster create response missing id")
				os.Exit(1)
			}
			clusterID = created.ID
			upsertResult = "created"
		} else {
			var updated clusterSummary
			if _, err := c.do(http.MethodPut, fmt.Sprintf("/api/clusters/%s", clusterID), clusterPayload, &updated); err != nil {
				fmt.Fprintln(os.Stderr, "ERROR:", err)
				os.Exit(1)
			}
			if updated.ID != "" {
				clusterID = updated.ID
			}
			upsertResult = "updated"
		}

		fmt.Println("cluster_upsert_result:", upsertResult)
		fmt.Println("cluster_id:", clusterID)

		if *testCluster {
			testVersion := ""
			err := retry("cluster connectivity test", *maxRetries, retryInterval, func() error {
				var resp clusterTestResponse
				status, err := c.do(
					http.MethodPost,
					fmt.Sprintf("/api/clusters/%s/test", clusterID),
					nil,
					&resp,
					http.StatusBadRequest,
				)
				if err != nil {
					return err
				}
				if status == http.StatusOK && resp.Success {
					testVersion = resp.Version
					return nil
				}
				if resp.Error != "" {
					return errors.New(resp.Error)
				}
				return errors.New("connectivity test returned success=false")
			})
			if err != nil {
				fmt.Fprintln(os.Stderr, "ERROR:", err)
				os.Exit(1)
			}
			fmt.Println("cluster_test_success:", true)
			fmt.Println("cluster_test_version:", testVersion)
		}
	}

	if !*skipDashboard {
		var dashboard dashboardSummary
		if _, err := c.do(http.MethodGet, "/api/dashboard", nil, &dashboard); err != nil {
			fmt.Fprintln(os.Stderr, "ERROR:", err)
			os.Exit(1)
		}
		fmt.Println("cluster_count:", dashboard.ClusterCount)
		fmt.Println("pod_count:", dashboard.PodCount)
		fmt.Println("deployment_count:", dashboard.DeploymentCount)
		fmt.Println("today_release_count:", dashboard.TodayReleaseCount)
	}

	var clusters []clusterSummary
	if _, err := c.do(http.MethodGet, "/api/clusters", nil, &clusters); err != nil {
		fmt.Fprintln(os.Stderr, "ERROR:", err)
		os.Exit(1)
	}
	fmt.Println("clusters_returned:", len(clusters))
}

File: scripts/python/launch_admin_via_api.py
#!/usr/bin/env python3
import argparse
import json
import os
import re
import sys
import time
import urllib.error
import urllib.request
from http.cookiejar import CookieJar
from pathlib import Path
from typing import Any, Callable


class AdminApiClient:
    def __init__(self, base_url: str):
        self.base_url = base_url.rstrip("/")
        self.cookies = CookieJar()
        self.opener = urllib.request.build_opener(urllib.request.HTTPCookieProcessor(self.cookies))

    def _request(self, method: str, path: str, payload: dict | None = None) -> tuple[int, Any]:
        url = f"{self.base_url}{path}"
        data = None
        headers = {"Content-Type": "application/json"}
        if payload is not None:
            data = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(url, method=method, data=data, headers=headers)
        try:
            with self.opener.open(req, timeout=20) as resp:
                body = resp.read().decode("utf-8")
                if not body:
                    return resp.status, {}
                try:
                    return resp.status, json.loads(body)
                except json.JSONDecodeError:
                    return resp.status, body
        except urllib.error.HTTPError as err:
            raw = err.read().decode("utf-8")
            try:
                parsed = json.loads(raw)
            except json.JSONDecodeError:
                parsed = raw
            return err.code, parsed
        except urllib.error.URLError as err:
            raise RuntimeError(f"request failed: {method} {path}: {err.reason}") from err

    def login(self, username: str, password: str) -> None:
        code, body = self._request("POST", "/api/auth/login", {"username": username, "password": password})
        if code != 200:
            raise RuntimeError(f"login failed: {code} {body}")

    def me(self) -> dict:
        code, body = self._request("GET", "/api/auth/me")
        if code != 200:
            raise RuntimeError(f"session validation failed: {code} {body}")
        if not isinstance(body, dict):
            raise RuntimeError("invalid /api/auth/me response type")
        return body

    def dashboard(self) -> dict:
        code, body = self._request("GET", "/api/dashboard")
        if code != 200:
            raise RuntimeError(f"dashboard request failed: {code} {body}")
        if not isinstance(body, dict):
            raise RuntimeError("invalid /api/dashboard response type")
        return body

    def clusters(self) -> list[dict]:
        code, body = self._request("GET", "/api/clusters")
        if code != 200:
            raise RuntimeError(f"clusters request failed: {code} {body}")
        if not isinstance(body, list):
            raise RuntimeError("invalid /api/clusters response type")
        return body

    def upsert_cluster(
        self,
        *,
        name: str,
        display_name: str,
        api_server_url: str,
        kubeconfig: str,
        description: str,
    ) -> tuple[str, str]:
        payload = {
            "displayName": display_name,
            "apiServerUrl": api_server_url,
            "authType": "kubeconfig",
            "kubeconfig": kubeconfig,
            "description": description,
            "notifyEnabled": False,
        }
        existing = next((cluster for cluster in self.clusters() if cluster.get("name") == name), None)
        if existing:
            cluster_id = existing.get("id")
            if not isinstance(cluster_id, str) or not cluster_id:
                raise RuntimeError(f"invalid existing cluster id: {cluster_id}")
            code, body = self._request("PUT", f"/api/clusters/{cluster_id}", payload)
            if code != 200:
                raise RuntimeError(f"cluster update failed: {code} {body}")
            return cluster_id, "updated"

        create_payload = {"name": name, **payload}
        code, body = self._request("POST", "/api/clusters", create_payload)
        if code != 201:
            raise RuntimeError(f"cluster create failed: {code} {body}")
        if not isinstance(body, dict) or not isinstance(body.get("id"), str):
            raise RuntimeError("cluster create response missing id")
        return body["id"], "created"

    def test_cluster(self, cluster_id: str) -> tuple[bool, dict]:
        code, body = self._request("POST", f"/api/clusters/{cluster_id}/test")
        if code == 200 and isinstance(body, dict) and body.get("success") is True:
            return True, body
        if code == 400:
            if isinstance(body, dict):
                return False, body
            return False, {"error": str(body)}
        raise RuntimeError(f"cluster test failed: {code} {body}")


def retry(action_name: str, max_retries: int, retry_interval_seconds: int, fn: Callable[[], Any]) -> Any:
    for attempt in range(1, max_retries + 1):
        try:
            return fn()
        except Exception as exc:  # noqa: BLE001
            if attempt == max_retries:
                raise RuntimeError(f"{action_name} failed after {attempt} attempts: {exc}") from exc
            print(
                f"{action_name} attempt {attempt}/{max_retries} failed: {exc}; retrying in {retry_interval_seconds}s",
                file=sys.stderr,
            )
            time.sleep(retry_interval_seconds)
    raise RuntimeError(f"{action_name} failed")


def extract_api_server_url(kubeconfig_content: str) -> str:
    match = re.search(r"(?m)^\s*server:\s*(\S+)\s*$", kubeconfig_content)
    if not match:
        raise RuntimeError("kubeconfig does not contain a server field")
    return match.group(1)


def resolve_password(password: str | None, password_env: str | None) -> str:
    if password:
        return password
    if password_env:
        env_password = os.getenv(password_env)
        if env_password:
            return env_password
        raise RuntimeError(f"environment variable {password_env} is empty")
    raise RuntimeError("password is required (--password or --password-env)")


def test_cluster_with_retry_error(client: AdminApiClient, cluster_id: str) -> dict:
    success, body = client.test_cluster(cluster_id)
    if success:
        return body
    raise RuntimeError(str(body.get("error", "connectivity test returned success=false")))


def main() -> int:
    parser = argparse.ArgumentParser(description="Launch/verify admin via API")
    parser.add_argument("--base-url", default="http://127.0.0.1:3002")
    parser.add_argument("--username", default="admin")
    parser.add_argument("--password")
    parser.add_argument("--password-env", default="")
    parser.add_argument("--upsert-cluster", action="store_true")
    parser.add_argument("--cluster-name")
    parser.add_argument("--cluster-display-name")
    parser.add_argument("--cluster-description", default="Managed by Terraform API workflow")
    parser.add_argument("--kubeconfig")
    parser.add_argument("--api-server-url", default="")
    parser.add_argument("--test-cluster", action="store_true")
    parser.add_argument("--max-retries", type=int, default=30)
    parser.add_argument("--retry-interval-seconds", type=int, default=2)
    parser.add_argument("--skip-dashboard", action="store_true")
    args = parser.parse_args()

    if args.max_retries <= 0:
        raise RuntimeError("--max-retries must be > 0")
    if args.retry_interval_seconds <= 0:
        raise RuntimeError("--retry-interval-seconds must be > 0")
    if args.upsert_cluster and (not args.cluster_name or not args.cluster_display_name or not args.kubeconfig):
        raise RuntimeError("--upsert-cluster requires --cluster-name, --cluster-display-name and --kubeconfig")

    password = resolve_password(args.password, args.password_env)
    client = AdminApiClient(args.base_url)
    retry("login", args.max_retries, args.retry_interval_seconds, lambda: client.login(args.username, password))
    user = client.me()

    print("login_user:", user.get("username"))

    if args.upsert_cluster:
        kubeconfig_content = Path(args.kubeconfig).read_text(encoding="utf-8")
        api_server_url = args.api_server_url or extract_api_server_url(kubeconfig_content)
        cluster_id, upsert_result = client.upsert_cluster(
            name=args.cluster_name,
            display_name=args.cluster_display_name,
            api_server_url=api_server_url,
            kubeconfig=kubeconfig_content,
            description=args.cluster_description,
        )
        print("cluster_upsert_result:", upsert_result)
        print("cluster_id:", cluster_id)

        if args.test_cluster:
            test_body = retry(
                "cluster connectivity test",
                args.max_retries,
                args.retry_interval_seconds,
                lambda: test_cluster_with_retry_error(client, cluster_id),
            )
            print("cluster_test_success:", test_body.get("success"))
            print("cluster_test_version:", test_body.get("version"))

    if not args.skip_dashboard:
        dashboard = client.dashboard()
        print("cluster_count:", dashboard.get("clusterCount"))
        print("pod_count:", dashboard.get("podCount"))
        print("deployment_count:", dashboard.get("deploymentCount"))
        print("today_release_count:", dashboard.get("todayReleaseCount"))

    clusters = client.clusters()
    print("clusters_returned:", len(clusters))
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        raise SystemExit(1)